source/common/common/logger.h

```diff
@@ -400,6 +401,35 @@ template <Id id> class Loggable {
     }                                                                                              \
   } while (0)
 
+#define ENVOY_LOG_FIRST_N(LEVEL, N, ...)                                                           \
+  do {                                                                                             \
+    static auto* countdown = new std::atomic<uint64_t>();                                          \
```

Comment:
This isn't thread-safe. If two threads do this at the same time, there could be a leak. I think this ends up being better as `static std::atomic<uint64_t> countdown;`. Either the compiler will see that it is initialized by zeroing, and do so at program start, or it will double-zero the same memory by initializing multiple times, but should avoid a leak at least.

Reply:
Wouldn't that lead up to a potential deinitialization fiasco? I was following the lead of https://github.com/envoyproxy/envoy/blob/f2306b35a73fe1c037c7006889ad31d8e54234cc/source/common/common/macros.h#L32
Reading up I am led to believe this would be thread safe since c++ 11 but I can't find a source that I would consider authorative,. Will dig some more.

Reply:
Relevant on the subject of leaking and stack vs heap allocation in the construct on first use idiom: [Why doesn’t the Construct On First Use Idiom use a static object instead of a static pointer?](https://isocpp.org/wiki/faq/ctors#construct-on-first-use-v2)

Reply:
https://en.cppreference.com/w/cpp/language/storage_duration#Static_local_variables has:

> If multiple threads attempt to initialize the same static local variable concurrently,  the initialization occurs exactly once (similar behavior can be obtained for arbitrary functions with std::call_once). Note: usual implementations of this feature use variants of the double-checked locking pattern,  which reduces runtime overhead for already-initialized local statics to a single non-atomic boolean comparison.

Reply:
Did some reading, and you're correct. TIL something new about static initialization.